File: ui/src/types/events.ts
export type EventType =
  | "agent_start"
  | "agent_end"
  | "step_start"
  | "step_end"
  | "tool_call"
  | "tool_result"
  | "llm_call"
  | "llm_response"
  | "state_update"
  | "error";

export type NodeKind = "agent" | "step" | "tool" | "llm";
export type NodeStatus = "pending" | "running" | "success" | "error";

export interface VapEvent {
  id: string;
  run_id: string;
  timestamp: number;
  type: EventType;
  node_id: string;
  node_kind: NodeKind;
  node_label: string;
  parent_id: string | null;
  data: Record<string, unknown>;
}

export interface GraphNode {
  id: string;
  kind: NodeKind;
  label: string;
  status: NodeStatus;
  parent_id: string | null;
  started_at: number | null;
  ended_at: number | null;
  data: Record<string, unknown>;
}

export interface GraphEdge {
  id: string;
  source: string;
  target: string;
  kind: string;
}

export interface RunGraph {
  run_id: string;
  label: string;
  app_id?: string | null; // set at run creation from agent_start data, not by the reducer
  status: NodeStatus;
  nodes: GraphNode[];
  edges: GraphEdge[];
  started_at: number;
  ended_at: number | null;
}

export interface RunSummary {
  run_id: string;
  label: string;
  app_id?: string | null; // stable pipeline id (trace(app_id=...)); grouping key = app_id ?? label
  status: NodeStatus;
  started_at: number;
  ended_at: number | null;
  node_count: number;
  event_count: number;
  total_cost_usd: number | null;
  tags: string[];
  summary?: string | null; // plain-language one-liner from vapviz/summary.py
}

// ── Cross-run analytics (GET /metrics) ──────────────────────────────────────

export interface TokenTotals {
  input: number;
  output: number;
}

export interface ModelStat {
  model: string;
  calls: number;
  cost_usd: number;
  input_tokens: number;
  output_tokens: number;
}

export interface KindCounts {
  agent: number;
  step: number;
  tool: number;
  llm: number;
}

export interface DailyCost {
  date: string;
  cost_usd: number;
  run_count: number;
}

export interface Metrics {
  run_count: number;
  success_count: number;
  error_count: number;
  running_count: number;
  success_rate: number | null;

  total_cost_usd: number;
  avg_cost_usd: number | null;

  total_duration_ms: number;
  avg_duration_ms: number | null;

  total_nodes: number;
  total_llm_calls: number;
  total_tokens: TokenTotals;

  by_model: ModelStat[];
  by_kind: KindCounts;
  cost_over_time: DailyCost[];
}

File: vap/events.py
from __future__ import annotations

from enum import Enum
from typing import Any, Optional
from pydantic import BaseModel, Field


class EventType(str, Enum):
    AGENT_START = "agent_start"
    AGENT_END = "agent_end"
    STEP_START = "step_start"
    STEP_END = "step_end"
    TOOL_CALL = "tool_call"
    TOOL_RESULT = "tool_result"
    LLM_CALL = "llm_call"
    LLM_RESPONSE = "llm_response"
    STATE_UPDATE = "state_update"
    ERROR = "error"


class NodeKind(str, Enum):
    AGENT = "agent"
    STEP = "step"
    TOOL = "tool"
    LLM = "llm"


class NodeStatus(str, Enum):
    PENDING = "pending"
    RUNNING = "running"
    SUCCESS = "success"
    ERROR = "error"


class VapEvent(BaseModel):
    id: str
    run_id: str
    timestamp: float
    type: EventType
    node_id: str
    node_kind: NodeKind
    node_label: str
    parent_id: Optional[str] = None
    data: dict[str, Any] = Field(default_factory=dict)
    schema_version: int = 1


class GraphNode(BaseModel):
    id: str
    kind: NodeKind
    label: str
    status: NodeStatus = NodeStatus.PENDING
    parent_id: Optional[str] = None
    started_at: Optional[float] = None
    ended_at: Optional[float] = None
    data: dict[str, Any] = Field(default_factory=dict)


class GraphEdge(BaseModel):
    id: str
    source: str
    target: str
    kind: str = "execution"


class RunSummary(BaseModel):
    run_id: str
    label: str
    status: NodeStatus
    started_at: float
    ended_at: Optional[float] = None
    node_count: int = 0
    event_count: int = 0
    total_cost_usd: Optional[float] = None


class RunGraph(BaseModel):
    run_id: str
    label: str
    status: NodeStatus
    nodes: list[GraphNode] = Field(default_factory=list)
    edges: list[GraphEdge] = Field(default_factory=list)
    started_at: float
    ended_at: Optional[float] = None
